[SPARK-32249][INFRA][2.4] Run Github Actions builds in branch-2.4

HyukjinKwon opened the pull request:
### What changes were proposed in this pull request?

This PR proposes to backport the following JIRAs:

- SPARK-32245
- SPARK-32292
- SPARK-32252
- SPARK-32408
- SPARK-32303
- SPARK-32363
- SPARK-32419
- SPARK-32491
- SPARK-32493
- SPARK-32496
- SPARK-32497
- SPARK-32357
- SPARK-32606
- SPARK-32605
- SPARK-32645
- Minor renaming https://github.com/apache/spark/commit/d0dfe4986b1c4cb5a47be46b2bbedeea42d81caf#diff-02d9c370a663741451423342d5869b21

in order to enable GitHub Actions in branch-2.4.

### Why are the changes needed?

To be able to run the tests in branch-2.4. Jenkins jobs are unstable.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

Build in this PR will test.

HyukjinKwon committed:
[SPARK-32245][INFRA] Run Spark tests in Github Actions

This PR aims to run the Spark tests in Github Actions.

To briefly explain the main idea:

- Reuse `dev/run-tests.py` with SBT build
- Reuse the modules in `dev/sparktestsupport/modules.py` to test each module
- Pass the modules to test into `dev/run-tests.py` directly via `TEST_ONLY_MODULES` environment variable. For example, `pyspark-sql,core,sql,hive`.
- `dev/run-tests.py` _does not_ take the dependent modules into account but solely the specified modules to test.

Another thing to note might be `SlowHiveTest` annotation. Running the tests in Hive modules takes too much so the slow tests are extracted and it runs as a separate job. It was extracted from the actual elapsed time in Jenkins:

![Screen Shot 2020-07-09 at 7 48 13 PM](https://user-images.githubusercontent.com/6477701/87050238-f6098e80-c238-11ea-9c4a-ab505af61381.png)

So, Hive tests are separated into to jobs. One is slow test cases, and the other one is the other test cases.

_Note that_ the current GitHub Actions build virtually copies what the default PR builder on Jenkins does (without other profiles such as JDK 11, Hadoop 2, etc.). The only exception is Kinesis https://github.com/apache/spark/pull/29057/files#diff-04eb107ee163a50b61281ca08f4e4c7bR23

Last week and onwards, the Jenkins machines became very unstable for many reasons:
  - Apparently, the machines became extremely slow. Almost all tests can't pass.
  - One machine (worker 4) started to have the corrupt `.m2` which fails the build.
  - Documentation build fails time to time for an unknown reason in Jenkins machine specifically. This is disabled for now at https://github.com/apache/spark/pull/29017.
  - Almost all PRs are basically blocked by this instability currently.

The advantages of using Github Actions:
  - To avoid depending on few persons who can access to the cluster.
  - To reduce the elapsed time in the build - we could split the tests (e.g., SQL, ML, CORE), and run them in parallel so the total build time will significantly reduce.
  - To control the environment more flexibly.
  - Other contributors can test and propose to fix Github Actions configurations so we can distribute this build management cost.

Note that:
- The current build in Jenkins takes _more than 7 hours_. With Github actions it takes _less than 2 hours_
- We can now control the environments especially for Python easily.
- The test and build look more stable than the Jenkins'.

No, dev-only change.

Tested at https://github.com/HyukjinKwon/spark/pull/4

Closes #29057 from HyukjinKwon/migrate-to-github-actions.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

dongjoon-hyun committed:
[SPARK-32245][INFRA][FOLLOWUP] Reenable Github Actions on commit

This PR reenables GitHub Action on every commit as a next step.

We carefully enabled GitHub Action on every PRs, and it looks good so far.

As we saw at https://github.com/apache/spark/pull/29072, GitHub Action is already triggered at every commits on every PRs. Enabling GitHub Action on `master` branch commit doesn't make a big difference. And, we need to start to test at every commit as a next step.

No.

Manual.

Closes #29076 from dongjoon-hyun/reenable_gha_commit.

Authored-by: Dongjoon Hyun <dongjoon@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

HyukjinKwon committed:
[SPARK-32292][SPARK-32252][INFRA] Run the relevant tests only in GitHub Actions

This PR mainly proposes to run only relevant tests just like Jenkins PR builder does. Currently, GitHub Actions always run full tests which wastes the resources.

In addition, this PR also fixes 3 more issues  very closely related together while I am here.

1. The main idea here is: It reuses the existing logic embedded in `dev/run-tests.py` which Jenkins PR builder use in order to run only the related test cases.

2. While I am here, I fixed SPARK-32292 too to run the doc tests. It was because other references were not available when it is cloned via `checkoutv2`. With `fetch-depth: 0`, the history is available.

3. In addition, it fixes the `dev/run-tests.py` to match with `python/run-tests.py` in terms of its options. Environment variables such as `TEST_ONLY_XXX` were moved as proper options. For example,

    ```bash
    dev/run-tests.py --modules sql,core
    ```

    which is consistent with `python/run-tests.py`, for example,

    ```bash
    python/run-tests.py --modules pyspark-core,pyspark-ml
    ```

4. Lastly, also fixed the formatting issue in module specification in the matrix:

    ```diff
    -            network_common, network_shuffle, repl, launcher
    +            network-common, network-shuffle, repl, launcher,
    ```

    which incorrectly runs build/test the modules.

By running only related tests, we can hugely save the resources and avoid unrelated flaky tests, etc.
Also, now it runs the doctest of `dev/run-tests.py` properly, the usages are similar between `dev/run-tests.py` and `python/run-tests.py`, and run `network-common`, `network-shuffle`, `launcher` and `examples` modules too.

No, dev-only.

Manually tested in my own forked Spark:

https://github.com/HyukjinKwon/spark/pull/7
https://github.com/HyukjinKwon/spark/pull/8
https://github.com/HyukjinKwon/spark/pull/9
https://github.com/HyukjinKwon/spark/pull/10
https://github.com/HyukjinKwon/spark/pull/11
https://github.com/HyukjinKwon/spark/pull/12

Closes #29086 from HyukjinKwon/SPARK-32292.

Authored-by: Hyukjin Kwon <gurwls223@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

HyukjinKwon committed:
[SPARK-32303][PYTHON][TESTS] Remove leftover from editable mode installation in PIP test

Currently the Jenkins PIP packaging test fails as below intermediately:

```
Installing dist into virtual env
Processing ./python/dist/pyspark-3.1.0.dev0.tar.gz
Collecting py4j==0.10.9 (from pyspark==3.1.0.dev0)
  Downloading https://files.pythonhosted.org/packages/9e/b6/6a4fb90cd235dc8e265a6a2067f2a2c99f0d91787f06aca4bcf7c23f3f80/py4j-0.10.9-py2.py3-none-any.whl (198kB)
Installing collected packages: py4j, pyspark
  Found existing installation: py4j 0.10.9
    Uninstalling py4j-0.10.9:
      Successfully uninstalled py4j-0.10.9
  Found existing installation: pyspark 3.1.0.dev0
Exception:
Traceback (most recent call last):
  File "/home/anaconda/envs/py36/lib/python3.6/site-packages/pip/_internal/cli/base_command.py", line 179, in main
    status = self.run(options, args)
  File "/home/anaconda/envs/py36/lib/python3.6/site-packages/pip/_internal/commands/install.py", line 393, in run
    use_user_site=options.use_user_site,
  File "/home/anaconda/envs/py36/lib/python3.6/site-packages/pip/_internal/req/__init__.py", line 50, in install_given_reqs
    auto_confirm=True
  File "/home/anaconda/envs/py36/lib/python3.6/site-packages/pip/_internal/req/req_install.py", line 816, in uninstall
    uninstalled_pathset = UninstallPathSet.from_dist(dist)
  File "/home/anaconda/envs/py36/lib/python3.6/site-packages/pip/_internal/req/req_uninstall.py", line 505, in from_dist
    '(at %s)' % (link_pointer, dist.project_name, dist.location)
AssertionError: Egg-link /home/jenkins/workspace/SparkPullRequestBuilder3/python does not match installed
```

- https://github.com/apache/spark/pull/29099#issuecomment-658073453 (amp-jenkins-worker-04)
- https://github.com/apache/spark/pull/29090#issuecomment-657819973 (amp-jenkins-worker-03)

Seems like the previous installation of editable mode affects other PRs.

This PR simply works around by removing the symbolic link from the previous editable installation. This is a common workaround up to my knowledge.

To recover the Jenkins build.

No, dev-only.

Jenkins build will test it out.

Closes #29102 from HyukjinKwon/SPARK-32303.

Lead-authored-by: HyukjinKwon <gurwls223@apache.org>
Co-authored-by: Hyukjin Kwon <gurwls223@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

HyukjinKwon committed:
[SPARK-32316][TESTS][INFRA] Test PySpark with Python 3.8 in Github Actions

This PR aims to test PySpark with Python 3.8 in Github Actions. In the script side, it is already ready:

https://github.com/apache/spark/blob/4ad9bfd53b84a6d2497668c73af6899bae14c187/python/run-tests.py#L161

This PR includes small related fixes together:

1. Install Python 3.8
2. Only install one Python implementation instead of installing many for SQL and Yarn test cases because they need one Python executable in their test cases that is higher than Python 2.
3. Do not install Python 2 which is not needed anymore after we dropped Python 2 at SPARK-32138
4. Remove a comment about installing PyPy3 on Jenkins - SPARK-32278. It is already installed.

Currently, only PyPy3 and Python 3.6 are being tested with PySpark in Github Actions. We should test the latest version of Python as well because some optimizations can be only enabled with Python 3.8+. See also https://github.com/apache/spark/pull/29114

No, dev-only.

Was not tested. Github Actions build in this PR will test it out.

Closes #29116 from HyukjinKwon/test-python3.8-togehter.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

HyukjinKwon committed:
[SPARK-32363][PYTHON][BUILD] Fix flakiness in pip package testing in Jenkins

This PR proposes:

- Don't use `--user` in pip packaging test
- Pull `source` out of the subshell, and place it first.
- Exclude user sitepackages in Python path during pip installation test

to address the flakiness of the pip packaging test in Jenkins.

(I think) #29116 caused this flakiness given my observation in the Jenkins log. I had to work around by specifying `--user` but it turned out that it does not properly work in old Conda on Jenkins for some reasons. Therefore, reverting this change back.

(I think) the installation at user site-packages affects other environments created by Conda in the old Conda version that Jenkins has. Seems it fails to isolate the environments for some reasons. So, it excludes user sitepackages in the Python path during the test.

In addition, #29116 also added some fallback logics of `conda (de)activate` and `source (de)activate` because Conda prefers to use `conda (de)activate` now per the official documentation and `source (de)activate` doesn't work for some reasons in certain environments (see also https://github.com/conda/conda/issues/7980). The problem was that `source` loads things to the current shell so does not affect the current shell. Therefore, this PR pulls `source` out of the subshell.

Disclaimer: I made the analysis purely based on Jenkins machine's log in this PR. It may have a different reason I missed during my observation.

To make the build and tests pass in Jenkins.

No, dev-only.

Jenkins tests should test it out.

Closes #29117 from HyukjinKwon/debug-conda.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

HyukjinKwon committed:
[SPARK-32408][BUILD] Enable crossPaths back to prevent side effects

### What changes were proposed in this pull request?

This PR proposes to enable `corssPaths` back for now to match with the build as it was.
It still indeterministically doesn't run JUnit tests given my observation, and this PR basically reverts the partial fix from https://github.com/apache/spark/pull/29057.

See also https://github.com/apache/spark/pull/29205 for the full context.

### Why are the changes needed?

To prevent the side effects from crossPaths such as SPARK_PREPEND_CLASSES or tests that run conditionally if the test classes are present in PySpark.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

Manually tested:

```bash
build/sbt -Phadoop-2.7 -Phive -Phive-2.3 -Phive-thriftserver -DskipTests clean test:package
./python/run-tests --python-executable=python3 --testname="pyspark.sql.tests.test_dataframe QueryExecutionListenerTests"
```

Closes #29218 from HyukjinKwon/SPARK-32408-1.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

HyukjinKwon committed:
[SPARK-32419][PYTHON][BUILD] Avoid using subshell for Conda env (de)activation in pip packaging test

### What changes were proposed in this pull request?

This PR proposes to avoid using subshell when it activates Conda environment. Looks like it ends up with activating the env within the subshell even if you use `conda` command.

### Why are the changes needed?

If you take a close look for GitHub Actions log:

```
 Installing dist into virtual env
Processing ./python/dist/pyspark-3.1.0.dev0.tar.gz
Collecting py4j==0.10.9
 Downloading py4j-0.10.9-py2.py3-none-any.whl (198 kB)
Using legacy setup.py install for pyspark, since package 'wheel' is not installed.
Installing collected packages: py4j, pyspark
 Running setup.py install for pyspark: started
 Running setup.py install for pyspark: finished with status 'done'
Successfully installed py4j-0.10.9 pyspark-3.1.0.dev0

...

Installing dist into virtual env
Obtaining file:///home/runner/work/spark/spark/python
Collecting py4j==0.10.9
 Downloading py4j-0.10.9-py2.py3-none-any.whl (198 kB)
Installing collected packages: py4j, pyspark
 Attempting uninstall: py4j
 Found existing installation: py4j 0.10.9
 Uninstalling py4j-0.10.9:
 Successfully uninstalled py4j-0.10.9
 Attempting uninstall: pyspark
 Found existing installation: pyspark 3.1.0.dev0
 Uninstalling pyspark-3.1.0.dev0:
 Successfully uninstalled pyspark-3.1.0.dev0
 Running setup.py develop for pyspark
Successfully installed py4j-0.10.9 pyspark
```

It looks not properly using Conda as it removes the previously installed one when it reinstalls again.
We should ideally test it with Conda environment as it's intended.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

GitHub Actions will test. I also manually tested in my local.

Closes #29212 from HyukjinKwon/SPARK-32419.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

HyukjinKwon committed:
[SPARK-32491][INFRA] Do not install SparkR in test-only mode in testing script

This PR proposes to skip SparkR installation that is to run R linters (see SPARK-8505) in the test-only mode at `dev/run-tests.py` script.

As of SPARK-32292, the test-only mode in `dev/run-tests.py` was introduced, for example:

```
dev/run-tests.py --modules sql,core
```

which only runs the relevant tests and does not run other tests such as linters. Therefore, we don't need to install SparkR when `--modules` are specified.

GitHub Actions build is currently failed as below:

```
ERROR: this R is version 3.4.4, package 'SparkR' requires R >= 3.5
[error] running /home/runner/work/spark/spark/R/install-dev.sh ; received return code 1
```

For some reasons, looks GitHub Actions started to have R 3.4.4 installed by default; however, R 3.4 was dropped as of SPARK-32073.  When SparkR tests are not needed, GitHub Actions still builds SparkR with a low R version and it causes the test failure.

This PR partially fixes it by avoid the installation of SparkR.

No, dev-only.

GitHub Actions tests should run to confirm this fix is correct.

Closes #29300 from HyukjinKwon/install-r.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

HyukjinKwon committed:
[SPARK-32493][INFRA] Manually install R instead of using setup-r in GitHub Actions

This PR proposes to manually install R instead of using `setup-r` which seems broken. Currently, GitHub Actions uses its default R 3.4.4 installed, which we dropped as of SPARK-32073.

While I am here, I am also upgrading R version to 4.0. Jenkins will test the old version and GitHub Actions tests the new version. AppVeyor uses R 4.0 but it does not check CRAN which is important when we make a release.

To recover GitHub Actions build.

No, dev-only

Manually tested at https://github.com/HyukjinKwon/spark/pull/15

Closes #29302 from HyukjinKwon/SPARK-32493.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

HyukjinKwon committed:
[SPARK-32496][INFRA] Include GitHub Action file as the changes in testing

### What changes were proposed in this pull request?

https://github.com/apache/spark/pull/26556 excluded `.github/workflows/master.yml`. So tests are skipped if the GitHub Actions configuration file is changed.

As of SPARK-32245, we now run the regular tests via the testing script. We should include it to test to make sure GitHub Actions build does not break due to some changes such as Python versions.

### Why are the changes needed?

For better test coverage in GitHub Actions build.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

GitHub Actions in this PR will test.

Closes #29305 from HyukjinKwon/SPARK-32496.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

HyukjinKwon committed:
[SPARK-32497][INFRA] Installs qpdf package for CRAN check in GitHub Actions

### What changes were proposed in this pull request?

CRAN check fails due to the size of the generated PDF docs as below:

```
...
 WARNING
‘qpdf’ is needed for checks on size reduction of PDFs
...
Status: 1 WARNING, 1 NOTE
See
  ‘/home/runner/work/spark/spark/R/SparkR.Rcheck/00check.log’
for details.
```

This PR proposes to install `qpdf` in GitHub Actions.

Note that I cannot reproduce in my local with the same R version so I am not documenting it for now.

Also, while I am here, I piggyback to install SparkR when the module includes `sparkr`. it is rather a followup of SPARK-32491.

### Why are the changes needed?

To fix SparkR CRAN check failure.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

GitHub Actions will test it out.

Closes #29306 from HyukjinKwon/SPARK-32497.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

HyukjinKwon committed:
[SPARK-32357][INFRA] Publish failed and succeeded test reports in GitHub Actions

This PR proposes to report the failed and succeeded tests in GitHub Actions in order to improve the development velocity by leveraging [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report). See the example below:

![Screen Shot 2020-08-13 at 8 17 52 PM](https://user-images.githubusercontent.com/6477701/90128649-28f7f280-dda2-11ea-9211-e98e34332f6b.png)

Note that we cannot just use [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report) in Apache Spark because PRs are from the forked repository, and GitHub secrets are unavailable for the security reason. This plugin and all similar plugins require to have the GitHub token that has the write access in order to post test results but it is unavailable in PRs.

To work around this limitation, I took this approach:

1. In workflow A, run the tests and upload the JUnit XML test results. GitHub provides to upload and download some files.
2. GitHub introduced new event type [`workflow_run`](https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/) 10 days ago. By leveraging this, it triggers another workflow B.
3. Workflow B is in the main repo instead of fork repo, and has the write access the plugin needs. In workflow B, it downloads the artifact uploaded from workflow A (from the forked repository).
4. Workflow B generates the test reports to port from JUnit xml files.
5. Workflow B looks up the PR and posts the test reports.

The `workflow_run` event is very new feature, and looks not so many GitHub Actions plugins support. In order to make this working with [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report), I had to fork two GitHub Actions plugins to use:
 - [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report) to have this custom fix: https://github.com/HyukjinKwon/action-surefire-report/commit/c96094cc35061fcf154a7cb46807f2f3e2339476
    It added `commit` argument to specify the commit to post the test reports. With `workflow_run`, it can access, in workflow B, to the commit from workflow A.

 - [dawidd6/action-download-artifact](https://github.com/dawidd6/action-download-artifact) to have this custom fix: https://github.com/HyukjinKwon/action-download-artifact/commit/750b71af351aba467757d7be6924199bb08db4ed
    It added the support of downloading all artifacts from workflow A, in workflow B. By default, it only supports to specify the name of artifact.

    Note that I was not able to use the official [actions/download-artifact](https://github.com/actions/download-artifact) because:
      - It does not support to download artifacts between different workflows, see also https://github.com/actions/download-artifact/issues/3. Once this issue is resolved, we can switch it back to [actions/download-artifact](https://github.com/actions/download-artifact).

I plan to make a pull request for both repositories so we don't have to rely on forks.

Currently, it's difficult to check the failed tests. You should scroll down long logs from GitHub Actions logs.

No, dev-only.

Manually tested at: https://github.com/HyukjinKwon/spark/pull/17, https://github.com/HyukjinKwon/spark/pull/18, https://github.com/HyukjinKwon/spark/pull/19, https://github.com/HyukjinKwon/spark/pull/20, and master branch of my forked repository.

Closes #29333 from HyukjinKwon/SPARK-32357-fix.

Lead-authored-by: Hyukjin Kwon <gurwls223@apache.org>
Co-authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

HyukjinKwon committed:
[SPARK-32606][SPARK-32605][INFRA] Remove the forks of action-surefire-report and action-download-artifact in test_report.yml

This PR proposes to remove the usage of my own forks and use the original plugins in GitHub Actions testing report.

SPARK-32357 introduced the GitHub Actions test reporting by leveraging two plugins:
 - [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report)
 - [dawidd6/action-download-artifact](https://github.com/dawidd6/action-download-artifact)

In order to make it working, it had to fork two repositories with custom fixes:
  - HyukjinKwon/action-surefire-reportc96094c
  - https://github.com/HyukjinKwon/action-download-artifact/commit/f86c565d52095e547b552eb46ba8e0696f6117f8

The two custom fixes are thankfully merged at https://github.com/ScaCap/action-surefire-report/pull/14 and https://github.com/dawidd6/action-download-artifact/pull/24, and they released new ones to use at [ScaCap/action-surefire-report/commits/v1](https://github.com/ScaCap/action-surefire-report/commits/v1) and [dawidd6/action-download-artifact/commits/v2](https://github.com/dawidd6/action-download-artifact/commits/v2)  - thanks jmisur and dawidd6 again.

To avoid relying on forks and code duplications.

No, dev-only.

Logically there is no diff. I tested it at https://github.com/HyukjinKwon/spark/runs/992824229 for doubly sure.

NOTE that this PR cannot be tested here within the workflow triggered by this PR without merging the changes in `test_report.yml` into the master.

Closes #29449 from HyukjinKwon/SPARK-32606-SPARK-32605.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

HyukjinKwon committed:
[MINOR][INFRA] Rename master.yml to build_and_test.yml

### What changes were proposed in this pull request?

This PR renames `master.yml` to `build_and_test.yml` to indicate this is the workflow that builds and runs the tests.

### Why are the changes needed?

Just for readability. `master.yml` looks like the name of the branch (to me).

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

GitHub Actions build in this PR will test it out.

Closes #29459 from HyukjinKwon/minor-rename.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: Gengliang Wang <gengliang.wang@databricks.com>

SparkQA commented:
**[Test build #127566 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127566/testReport)** for PR 29465 at commit [`86e2ccf`](https://github.com/apache/spark/commit/86e2ccf09ba8a4b5cec2bdf4082ae344267717ca).

SparkQA commented:
**[Test build #127566 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127566/testReport)** for PR 29465 at commit [`86e2ccf`](https://github.com/apache/spark/commit/86e2ccf09ba8a4b5cec2bdf4082ae344267717ca).
 * This patch **fails due to an unknown error code, 255**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127566/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32189/
Test PASSed.

SparkQA commented:
**[Test build #127568 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127568/testReport)** for PR 29465 at commit [`ea8c487`](https://github.com/apache/spark/commit/ea8c4876f21390fc0ded350409351589feb6171b).

HyukjinKwon commented on .github/workflows/master.yml:
Hadoop 2.6 is default.

HyukjinKwon commented on .github/workflows/master.yml:
Hive 2.3 does not exist

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32191/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32192/
Test PASSed.

HyukjinKwon commented on .github/workflows/master.yml:
streaming-flume, streaming-flume-sink and streaming-kafka-0-8 exist

HyukjinKwon commented on .github/workflows/master.yml:
Python 2.7 is tested in branch-2.4.

HyukjinKwon commented on .github/workflows/master.yml:
JDK 11 is not supported yet.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32193/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127568/
Test FAILed.

SparkQA commented:
**[Test build #127569 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127569/testReport)** for PR 29465 at commit [`a27f7d6`](https://github.com/apache/spark/commit/a27f7d667c68b7f0f813b1e59f1474c78763dcee).

SparkQA commented:
**[Test build #127570 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127570/testReport)** for PR 29465 at commit [`69f2509`](https://github.com/apache/spark/commit/69f2509a4b9fbb8466dd952ea2e9473d93c9e889).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32194/
Test PASSed.

SparkQA commented:
**[Test build #127571 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127571/testReport)** for PR 29465 at commit [`c043e0e`](https://github.com/apache/spark/commit/c043e0eab2d4498d61350276cdb24e85245307f4).

SparkQA commented:
**[Test build #127570 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127570/testReport)** for PR 29465 at commit [`69f2509`](https://github.com/apache/spark/commit/69f2509a4b9fbb8466dd952ea2e9473d93c9e889).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127570/
Test FAILed.

SparkQA commented:
**[Test build #127571 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127571/testReport)** for PR 29465 at commit [`c043e0e`](https://github.com/apache/spark/commit/c043e0eab2d4498d61350276cdb24e85245307f4).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127571/
Test FAILed.

SparkQA commented:
**[Test build #127572 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127572/testReport)** for PR 29465 at commit [`4400c16`](https://github.com/apache/spark/commit/4400c16c9c1183deba2be3480b55355fcffc510b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32195/
Test PASSed.

SparkQA commented:
**[Test build #127572 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127572/testReport)** for PR 29465 at commit [`4400c16`](https://github.com/apache/spark/commit/4400c16c9c1183deba2be3480b55355fcffc510b).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127572/
Test FAILed.

HyukjinKwon commented on .github/workflows/master.yml:
In branch-2.4, python 2 is installed last so the default `python` is Python 2.
This is because we use `python` shebang in most of scripts, and Python 3 was not supported at that time in these scripts.

HyukjinKwon commented:
FYI @dongjoon-hyun, @viirya, @gengliangwang.

SparkQA commented:
**[Test build #127575 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127575/testReport)** for PR 29465 at commit [`b57b290`](https://github.com/apache/spark/commit/b57b29083db167212a6abcc36607ab1572d2a1c2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32198/
Test PASSed.

SparkQA commented:
**[Test build #127575 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127575/testReport)** for PR 29465 at commit [`b57b290`](https://github.com/apache/spark/commit/b57b29083db167212a6abcc36607ab1572d2a1c2).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127575/
Test FAILed.

SparkQA commented:
**[Test build #127576 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127576/testReport)** for PR 29465 at commit [`96331e6`](https://github.com/apache/spark/commit/96331e62822b5906158d77fd07f84211117b20c1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32199/
Test PASSed.

HyukjinKwon commented on .github/workflows/master.yml:
`org.apache.spark.tags.ExtendedSQLTest` does not exist in branch-2.4 (SPARK-29191). So, SQL tests are not split in branch-2.4. It's split in branch-3.0 and master.

SparkQA commented:
**[Test build #127578 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127578/testReport)** for PR 29465 at commit [`7e6efd9`](https://github.com/apache/spark/commit/7e6efd90a3c69c4a1b7321a1e24b432f63954efe).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32201/
Test PASSed.

SparkQA commented:
**[Test build #127569 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127569/testReport)** for PR 29465 at commit [`a27f7d6`](https://github.com/apache/spark/commit/a27f7d667c68b7f0f813b1e59f1474c78763dcee).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127569/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #127580 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127580/testReport)** for PR 29465 at commit [`7e6efd9`](https://github.com/apache/spark/commit/7e6efd90a3c69c4a1b7321a1e24b432f63954efe).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32203/
Test PASSed.

SparkQA commented:
**[Test build #127576 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127576/testReport)** for PR 29465 at commit [`96331e6`](https://github.com/apache/spark/commit/96331e62822b5906158d77fd07f84211117b20c1).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127576/
Test FAILed.

SparkQA commented:
**[Test build #127578 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127578/testReport)** for PR 29465 at commit [`7e6efd9`](https://github.com/apache/spark/commit/7e6efd90a3c69c4a1b7321a1e24b432f63954efe).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127578/
Test FAILed.

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32207/
Test PASSed.

SparkQA commented:
**[Test build #127584 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127584/testReport)** for PR 29465 at commit [`7e6efd9`](https://github.com/apache/spark/commit/7e6efd90a3c69c4a1b7321a1e24b432f63954efe).

HyukjinKwon commented on .github/workflows/master.yml:
branch-2.4 does not support Python 3.8 due to the old cloudpickle (SPARK-29536).

SparkQA commented:
**[Test build #127580 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127580/testReport)** for PR 29465 at commit [`7e6efd9`](https://github.com/apache/spark/commit/7e6efd90a3c69c4a1b7321a1e24b432f63954efe).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #127585 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127585/testReport)** for PR 29465 at commit [`df84636`](https://github.com/apache/spark/commit/df846368ccfb2d98b272f6f67c7b4ba87656fad8).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127580/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32208/
Test PASSed.

SparkQA commented:
**[Test build #127585 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127585/testReport)** for PR 29465 at commit [`df84636`](https://github.com/apache/spark/commit/df846368ccfb2d98b272f6f67c7b4ba87656fad8).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127585/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127584/
Test FAILed.

SparkQA commented:
**[Test build #127586 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127586/testReport)** for PR 29465 at commit [`500b534`](https://github.com/apache/spark/commit/500b534a402d69f320dc983e140174d7708bc92e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32209/
Test PASSed.

SparkQA commented:
**[Test build #127586 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127586/testReport)** for PR 29465 at commit [`500b534`](https://github.com/apache/spark/commit/500b534a402d69f320dc983e140174d7708bc92e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127586/
Test PASSed.

SparkQA commented:
**[Test build #127604 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127604/testReport)** for PR 29465 at commit [`358be33`](https://github.com/apache/spark/commit/358be336ed2adb2521ca61dd5345ce71a8030051).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32227/
Test PASSed.

HyukjinKwon commented on .github/workflows/master.yml:
As of SPARK-28752, `pygments.rb` was removed but it still needs in branch-2.4.

HyukjinKwon commented:
Now it should pass all tests. I will port back once I get a green

SparkQA commented:
**[Test build #127608 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127608/testReport)** for PR 29465 at commit [`ed33409`](https://github.com/apache/spark/commit/ed334099b0ffd5fffa846e7651f706a659d29f1c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32231/
Test PASSed.

SparkQA commented:
**[Test build #127604 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127604/testReport)** for PR 29465 at commit [`358be33`](https://github.com/apache/spark/commit/358be336ed2adb2521ca61dd5345ce71a8030051).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127604/
Test PASSed.

SparkQA commented:
**[Test build #127611 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127611/testReport)** for PR 29465 at commit [`a74e610`](https://github.com/apache/spark/commit/a74e610fd407e727ed70fe3c32bc76788f599f97).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32234/
Test PASSed.

SparkQA commented:
**[Test build #127611 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127611/testReport)** for PR 29465 at commit [`a74e610`](https://github.com/apache/spark/commit/a74e610fd407e727ed70fe3c32bc76788f599f97).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127611/
Test FAILed.

SparkQA commented:
**[Test build #127613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127613/testReport)** for PR 29465 at commit [`a0a7351`](https://github.com/apache/spark/commit/a0a73518137e46be0ae45218fbd3a7cb6b6d5712).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32236/
Test PASSed.

HyukjinKwon committed:
[SPARK-32645][INFRA] Upload unit-tests.log as an artifact

### What changes were proposed in this pull request?

This PR proposes to upload `target/unit-tests.log` into the artifact so it will be able to download here:
![Screen Shot 2020-08-18 at 2 23 18 PM](https://user-images.githubusercontent.com/6477701/90474095-789e3b80-e15f-11ea-87f8-e7da3df3c03e.png)

### Why are the changes needed?

Jenkins has this feature. It should be best to have the same dev functionalities with it.
Also, note that this was pointed out https://github.com/apache/spark/pull/29225#discussion_r471485011.

### Does this PR introduce _any_ user-facing change?

No, dev-only

### How was this patch tested?

https://github.com/apache/spark/actions/runs/213000777 should demonstrate it

Closes #29454 from HyukjinKwon/SPARK-32645.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

HyukjinKwon commented:
It should be ready to be reviewed or merged now.

SparkQA commented:
**[Test build #127615 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127615/testReport)** for PR 29465 at commit [`1393174`](https://github.com/apache/spark/commit/1393174703bdc7236d2c96ae9efdbaa8f8993f85).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32238/
Test PASSed.

SparkQA commented:
**[Test build #127608 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127608/testReport)** for PR 29465 at commit [`ed33409`](https://github.com/apache/spark/commit/ed334099b0ffd5fffa846e7651f706a659d29f1c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #127615 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127615/testReport)** for PR 29465 at commit [`1393174`](https://github.com/apache/spark/commit/1393174703bdc7236d2c96ae9efdbaa8f8993f85).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127615/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127608/
Test PASSed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #127617 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127617/testReport)** for PR 29465 at commit [`1393174`](https://github.com/apache/spark/commit/1393174703bdc7236d2c96ae9efdbaa8f8993f85).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32241/
Test PASSed.

HyukjinKwon commented:
It should be ready to be reviewed or merged now.

HyukjinKwon commented on .github/workflows/build_and_test.yml:
Some tests do not pass in PySpark with PyArrow, for example, `pyspark.sql.tests.ArrowTests` with Python 2.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127617/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127613/
Test FAILed.

HyukjinKwon commented on .github/workflows/build_and_test.yml:
It's not GitHub Actions specific. Jenkins does not test Python 2 with PyArrow and I can reproduce it in my local as well.

SparkQA commented:
**[Test build #127622 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127622/testReport)** for PR 29465 at commit [`84846a8`](https://github.com/apache/spark/commit/84846a873658fa609148e208f403d2e010e4114b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32246/
Test PASSed.

SparkQA commented:
**[Test build #127622 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127622/testReport)** for PR 29465 at commit [`84846a8`](https://github.com/apache/spark/commit/84846a873658fa609148e208f403d2e010e4114b).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127622/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #127633 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127633/testReport)** for PR 29465 at commit [`84846a8`](https://github.com/apache/spark/commit/84846a873658fa609148e208f403d2e010e4114b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32257/
Test PASSed.

SparkQA commented:
**[Test build #127633 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127633/testReport)** for PR 29465 at commit [`84846a8`](https://github.com/apache/spark/commit/84846a873658fa609148e208f403d2e010e4114b).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127633/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #127638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127638/testReport)** for PR 29465 at commit [`84846a8`](https://github.com/apache/spark/commit/84846a873658fa609148e208f403d2e010e4114b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32262/
Test PASSed.

SparkQA commented:
**[Test build #127638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127638/testReport)** for PR 29465 at commit [`84846a8`](https://github.com/apache/spark/commit/84846a873658fa609148e208f403d2e010e4114b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127638/
Test PASSed.

HyukjinKwon commented:
The test failure in PySpark ML will be addressed at https://github.com/apache/spark/pull/29481.

@ScrapCodes and @dongjoon-hyun, can you review this and #29481? I think it's ready to go.

dongjoon-hyun reviewed (approved):
Sure. +1, LGTM. Merged to branch-2.4

HyukjinKwon commented:
Thank you @dongjoon-hyun!
